packages/plugin-ext/src/plugin/scm.ts

```diff
@@ -0,0 +1,347 @@
+        }
+    }
+
+    get statusBarCommands(): theia.Command[] | undefined {
+        return this._statusBarCommands;
+    }
+
+    set statusBarCommands(statusBarCommands: theia.Command[] | undefined) {
+        this._statusBarCommands = statusBarCommands;
+        if (statusBarCommands) {
+            const commands = statusBarCommands.map(statusBarCommand => {
+                const command = {
+                    id: statusBarCommand.command ? statusBarCommand.command : '',
+                    text: statusBarCommand.title ? statusBarCommand.title : '',
+                    alignment: 0
+                };
+                return command;
+            });
+            this.proxy.$updateSourceControl(this.handle, {statusBarCommands: commands});
+        }
+    }
+
+    getResourceGroup(handle: number): theia.SourceControlResourceGroup | undefined {
+        return this.resourceGroupsMap.get(handle);
+    }
+}
+
+class SourceControlResourceGroupImpl implements theia.SourceControlResourceGroup {
+
+    private static handle: number = 0;
+    private static resourceHandle: number = 0;
+    private handle = SourceControlResourceGroupImpl.handle ++;
+    private _hideWhenEmpty: boolean | undefined = undefined;
+    private _resourceStates: theia.SourceControlResourceState[] = [];
+    private resourceStatesMap: Map<number, theia.SourceControlResourceState> = new Map();
+
+    constructor(
+        private proxy: ScmMain,
+        private commands: CommandRegistryExt,
+        private sourceControlHandle: number,
+        private _id: string,
+        private _label: string,
+    ) {
+        this.proxy.$registerGroup(sourceControlHandle, this.handle, _id, _label);
+    }
+
+    get id(): string {
+        return this._id;
+    }
+
+    get label(): string {
+        return this._label;
+    }
+
+    set label(label: string) {
+        this._label = label;
+        this.proxy.$updateGroupLabel(this.sourceControlHandle, this.handle, label);
+    }
+
+    get hideWhenEmpty(): boolean | undefined {
+        return this._hideWhenEmpty;
+    }
+
+    set hideWhenEmpty(hideWhenEmpty: boolean | undefined) {
+        this._hideWhenEmpty = hideWhenEmpty;
+        this.proxy.$updateGroup(this.sourceControlHandle, this.handle, {hideWhenEmpty});
+    }
+
+    get resourceStates(): theia.SourceControlResourceState[] {
+        return this._resourceStates;
+    }
+
+    set resourceStates(resources: theia.SourceControlResourceState[]) {
+        this._resourceStates = resources;
+        this.resourceStatesMap.clear();
+        this.proxy.$updateResourceState(this.sourceControlHandle, this.handle, resources.map(resourceState => {
+            const handle = SourceControlResourceGroupImpl.resourceHandle ++;
+            let command;
+            let decorations;
+            if (resourceState.command) {
+                const { id, title, tooltip } = resourceState.command;
```

Comment:
`id` is deprecated use `command` instead

Reply:
fixed